datafusion/core/src/physical_plan/aggregates/row_hash.rs

```diff
@@ -270,10 +315,25 @@ fn group_aggregate_batch(
                 // Existing entry for this group value
                 Some((_hash, group_idx)) => {
                     let group_state = &mut group_states[*group_idx];
+
                     // 1.3
                     if group_state.indices.is_empty() {
                         groups_with_rows.push(*group_idx);
                     };
+
+                    // ensure we have enough indices allocated
+                    if group_state.indices.capacity() == group_state.indices.len() {
+                        // allocate more
+
+                        // growth factor: 2, but at least 2 elements
+                        let bump_elements = (group_state.indices.capacity() * 2).max(2);
```

Comment:
I wonder if we could somehow encapsulate the memory manager interactions into functions on `GroupAggrState` rather than treating it like a struct. I don't think that is necessary .

However encapsulating might:
1. Keep this code manageable for future readers
2. Allow the memory allocation routines to be unit tested (like that when new groups are added that the memory allocation is incremented correctly)

Reply:
I tend to agree with with @alamb here, IMHO `group_aggregate_batch` is too busy at the moment, and some kind of separation of concerns would help. 

What if `group_aggregate_batch` returns how much more memory it allocated, and  accounting is done after method call? This would help with encapsulation of aggregation algorithm and make it easier to swap. I'm aware that it might not produce 100% correct results but as we discussed in #3941 it is ok to have small discrepancy for short period of time

Reply:
Also, this way end of the batch would be a "safe point" at which we could trigger spill

Reply:
done.

Let me know if this looks better. I will pull out + document all the helper structs and traits when I port V1 (I want at least a 2nd consumer so I can make sure the interface makes sense).